outrank/algorithms/synthetic_data_generators/cc_generator.py

```diff
@@ -0,0 +1,849 @@
+        :param random_values: flag, enables random (integer) feature values from set [low, high]
+        :param low: sets lower bound of random feature values
+        :param high: sets high bound of random feature values
+        :param seed: sets seed of numpy random
+        :return: X, 2D dataset
+        """
+
+        self.dataset_info.update({
+            'general': {
+                'n_features': n_features,
+                'n_samples': n_samples,
+                'cardinality': cardinality,
+                'structure': structure,
+                'ensure_rep': ensure_rep,
+                'seed': seed,
+            },
+        })
+
+        np.random.seed(seed)
+        X = np.empty([n_features, n_samples])
+
+        if structure is None:
+            # No specific structure parameter passed
+            for i in range(n_features):
+                x = self._generate_feature(
+                    n_samples,
+                    cardinality=cardinality,
+                    ensure_rep=ensure_rep,
+                    random_values=random_values,
+                    low=low,
+                    high=high,
+                )
+                X[i] = x
+        else:
+            # Structure parameter passed, building based on structure
+            ix = 0
+            for data in structure:
+                if not isinstance(data[0], (list, np.ndarray)):
+                    # Data in structure is a tuple of (feature index (integer), feature attributes)
+                    feature_ix, feature_attributes = data
+
+                    if ix < feature_ix:
+                        # Filling out the dataset up to column index feature_ix
+                        for i in range(ix, feature_ix):
+                            x = self._generate_feature(
+                                n_samples,
+                                cardinality=cardinality,
+                                ensure_rep=ensure_rep,
+                                random_values=random_values,
+                                low=low,
+                                high=high,
+                            )
+                            X[ix] = x
+                            ix += 1
+
+                    x = self._configure_generate_feature(
+                        feature_attributes,
+                        n_samples,
+                        ensure_rep=ensure_rep,
+                        random_values=random_values,
+                        low=low,
+                        high=high,
+                    )
+                    X[ix] = x
+                    ix += 1
+
+                else:
+                    # Data in structure is a tuple of (list of feature indexes, feature attributes)
+                    feature_ixs, feature_attributes = data
+
+                    for feature_ix in feature_ixs:
+                        # Filling out the dataset up to feature_ix
+                        if ix < feature_ix:
```

Comment:
Also a general comment - it's not math source, you can name variables with fuller/more idiomatic names, not this PR though